src/main/java/com/google/cloud/spanner/myadapter/session/SessionState.java

```diff
@@ -29,4 +82,100 @@ public ProtocolStatus getProtocolStatus() {
   public void setProtocolStatus(ProtocolStatus protocolStatus) {
     this.protocolStatus = protocolStatus;
   }
+
+  Map<String, SystemVariable> getVariableMapForScope(SessionVariableScope scope) {
+    switch (scope) {
+      case GLOBAL:
+        return GLOBAL_SETTINGS;
+      case SESSION:
+        return settings;
+      default:
+        throw unknownParamError("scope");
+    }
+  }
+  /**
+   * Sets the value of the specified setting. The new value will be persisted if the current
+   * transaction is committed. The value will be lost if the transaction is rolled back.
+   */
+  public void set(String name, String value, SessionVariableScope scope) {
+    logger.log(
+        Level.INFO,
+        () -> String.format("Setting system variable %s to %s at scope %s", name, value, scope));
+    Map<String, SystemVariable> variableMap = getVariableMapForScope(scope);
+    internalSet(name.toLowerCase(Locale.ROOT), value, variableMap);
+  }
+
+  private void internalSet(String name, String value, Map<String, SystemVariable> variableMap) {
+    if ("names".equals(name)) {
+      // TODO: Consider handling "SET NAMES" as a separate statement type.
+      handleNames(value);
+      return;
+    }
+    if ("autocommit".equals(name)) {
+      // Autocommit value needs to be converted to an integer as internally autocommit is being
+      // tracked as integer.
+      value = handleAutocommit(value);
+    }
+    SystemVariable variable = variableMap.get(name);
+    if (variable == null) {
+      throw unknownVariableError(name);
+    }
+    variable.setValue(value);
+  }
+
+  private String handleAutocommit(String value) {
+    if (BooleanParser.TRUE_VALUES.contains(value)) {
+      backendConnection.processSetAutocommit();
```

Comment:
Instead on SessionState calling backend connection, design the method to return the operation. Otherwise session state will become very bloted. Also, this introduces circular dependency which is an Anti-pattern.

Reply:
I get your point, but I couldn't think of a generic way of "returning the operation" here.  So I have tried to solved it using a patten of having a pre and post hooks(or handlers) for an operation. Please check the latest commit addressing the comments.